### PR TITLE
Feat(internal): Store Fleets in a class for later use.

### PR DESCRIPTION
**Pull request by RisingLeaf**
**Feature:** This PR implements the feature request detailed and discussed on discord.

## Feature Details
Renames fleet class to fleetloader and creates a new fleet class that stores an instance of a fleet

## Testing Done
Not really something to test here.

## Performance Impact
N/A

### CODE REVIEW

File: source/Engine.cpp [hunk 1]
```diff
@@ -1596,6 +1595,20 @@ void Engine::CalculateStep()
...
+	{
+		bool shouldBeRemoved = true;
+		for(shared_ptr<Ship> ship : fleet.ships)
+			if(ship.get())
+				shouldBeRemobed = false;
+		if(shouldBeRemove)
+			it = fleets.erase(it);
+		else
+			++it;
+	}
```
Comment (RisingLeaf) [inline source/Engine.cpp:1610]:
```suggestion
	auto it = fleets.begin();
	while(it != fleets.end())
	{
		bool shouldBeRemoved = true;
		for(shared_ptr<Ship> ship : fleet.ships)
			if(ship.get())
				shouldBeRemoved = false;
		if(shouldBeRemoved)
			it = fleets.erase(it);
		else
			++it;
	}
```

File: source/Engine.cpp [hunk 2]
```diff
@@ -1596,6 +1595,20 @@ void Engine::CalculateStep()
...
+	{
+		bool shouldBeRemoved = true;
+		for(shared_ptr<Ship> ship : fleet.ships)
+			if(ship.get())
+				shouldBeRemoved = false;
+		if(shouldBeRemoved)
+			it = fleets.erase(it);
+		else
+			++it;
+	}
```
Comment (UnorderedSigh) [inline source/Engine.cpp:1610]: This compiles on ubuntu 18 with g++

```suggestion
	for(auto it = fleets.begin(); it != fleets.end(); )
	{
		bool shouldBeRemoved = true;
		for(shared_ptr<Ship> ship : it->ships)
			if(ship.get())
			{
				shouldBeRemoved = false;
				break;
			}
		if(shouldBeRemoved)
			it = fleets.erase(it);
		else
			++it;
	}
```

File: source/Fleet.cpp [hunk 3]
```diff
@@ -436,9 +440,70 @@ void Fleet::Enter(const System &system, list<shared_ptr<Ship>> &ships, const Pla
...
+
+		SetCargo(&*ship);
+	}
+}
+
+
+
+// Place one of the variants in the given system, already "in action." If the carried flag is set,
+// only uncarried ships will be added to the list (as any carriables will be stored in bays).
+void Fleet::Place(const System &system, list<shared_ptr<Ship>> &ships, vector<FleetHolder> &fleets, bool carried) const
```
Comment (RisingLeaf) [inline source/Fleet.cpp:458]:
```suggestion
void Fleet::Place(const System &system, list<shared_ptr<Ship>> &ships, vector<FleetHolder> &fleets,
	bool carried) const
```

File: source/Fleet.cpp [hunk 4]
```diff
@@ -277,7 +277,7 @@ const Government *Fleet::GetGovernment() const
 
 
 // Choose a fleet to be created during flight, and have it enter the system via jump or planetary departure.
-void Fleet::Enter(const System &system, list<shared_ptr<Ship>> &ships, const Planet *planet) const
+void Fleet::Enter(const System &system, list<shared_ptr<Ship>> &ships, vector<FleetHolder> &fleets, const Planet *planet) const
```
Comment (RisingLeaf) [inline source/Fleet.cpp:280]:
```suggestion
void Fleet::Enter(const System &system, list<shared_ptr<Ship>> &ships, vector<FleetHolder> &fleets,
	const Planet *planet) const
```

**Review by Amazinite — CHANGES_REQUESTED** [review]
I don't think that this "FleetHolder" should simply be a struct contained within Fleet, as that doesn't give us very much room for customization. Structs don't have functions, so anything we do to modify a fleet would need to be done by a separate class.

I think that instead we should move toward having a "FleetBuilder -> Fleet" pattern, similar to what can be seen in #6502 by splitting our currently massive "Ship" class into "ShipLoader" and "Ship" where ShipLoaderis what holds Ship definitions and instantiates Ships. Another example of this can be seen in Hazard and Weather; Hazard contains our definition, but when a Hazard is actually in effect in game, it becomes instantiated into a Weather.

So rename our current Fleet class to FleetBuilder (or maybe FleetLoader to match #6502), rename FleetHolder to just Fleet and make it its own class, and have FleetBuilder create Fleets. Instead of the vector of Fleets in Engine then being passed to the FleetBuilder, the FleetBuilder should return a Fleet when Place is called, with that Fleet being emplaced into the vector. That is, instead of
```
fleet.Get()->Place(*system, newShips, fleets);
```
we'd have
```
fleets.emplace_back(fleet.Get()->Place(*system, newShips));
```
And this might not be the case at first, but in the future we might also be able to get rid of passing `newShips` to the FleetBuilder, since each Fleet in Engine's vector will contain all of the current ships.

For now, Fleet should probably just have Step and ShouldBeRemoved functions so that we can use Prune on Fleet. Stepping Fleet would just involve determining if ShouldBeRemoved should return true (if all the Fleet's Ships have been pruned, as you're currently doing).

Even just these basic changes are probably going to be a lot more than what you've done with making this struct, but we'll have a much more solid foundation for future changes this way.

**Comment by RisingLeaf** [comment]
> I don't think that this "FleetHolder" should simply be a struct contained within Fleet, as that doesn't give us very much room for customization. Structs don't have functions, so anything we do to modify a fleet would need to be done by a separate class.

It is new to me that structs can not have functions, but I see your point, I will implemented the suggested changes soon(hopefully this weekend)

**Comment by Amazinite** [comment]
> It is new to me that structs can not have functions

I actually looked into it and I'm wrong on that. It's just that when you use a struct, it's usually because you're just packaging data/variables (which is made slightly easier by the contents of structs being public by default, whereas class contents are private by default), so you typically don't have functions. Although it looks like structs can't have constructors, which we would probably want in this case.

File: source/Fleet.h [hunk 5]
```diff
@@ -45,6 +45,14 @@ class System;
 // AI personality, and set of friendly and hostile "hail" messages, and the ship
 // names are chosen based on a given random "phrase" generator.
 class Fleet {
+public:
+	// Struct to hold an instance of a fleet.
+	struct FleetHolder {
+		std::string name;
+		std::shared_ptr<Ship> flagship;
+		std::set<std::shared_ptr<Ship>> ships;
```
Comment (UnorderedSigh) [inline source/Fleet.h:53]: This should be a vector or list, so you don't lose the information about the ordering of the ships in the fleet. Given that the number is small, a vector is probably the fastest.

Also, `shared_ptr` comparisons use the pointer value, not the object, so using an ordered set is pointless. If you do want to discard the information about the ordering, then you should use an unordered_set instead. That'll also get rid of the tree management that is usually more cost than it is worth when the number of objects is small.

File: source/Fleet.h [hunk 6]
```diff
@@ -61,9 +69,12 @@ class Fleet {
 	const Government *GetGovernment() const;
 
 	// Choose a fleet to be created during flight, and have it enter the system via jump or planetary departure.
-	void Enter(const System &system, std::list<std::shared_ptr<Ship>> &ships, const Planet *planet = nullptr) const;
+	void Enter(const System &system, std::list<std::shared_ptr<Ship>> &ships, std::vector<FleetHolder> &fleets,
```
Comment (UnorderedSigh) [inline source/Fleet.h:72]: Why does the Fleet need the internal state of the Engine? If you're putting a spawned fleet into a container, can't you just return the container? (Or a shared_ptr to it?)

Comment (RisingLeaf) [reply source/Fleet.h]: done, somehow, great overhaul

File: source/Fleet.h (hunk 5 above)
Comment (RisingLeaf) [reply source/Fleet.h]: done

**Pull request closed without merging**